profiler/src/ProfilerEngine/Datadog.Profiler.Native/StackSamplerLoop.cpp

```diff
@@ -165,20 +175,46 @@ void StackSamplerLoop::MainLoop()
 
 void StackSamplerLoop::MainLoopIteration()
 {
+    int64_t timestampNanosecs1 = 0;
+    int64_t timestampNanosecs2 = 0;
+
     // In each iteration, a few threads are sampled to compute wall time.
-    if (_pConfiguration->IsWallTimeProfilingEnabled())
+    if (_isWalltimeEnabled)
     {
+        if (_areInternalMetricsEnabled)
+        {
+            timestampNanosecs1 = OpSysTools::GetHighPrecisionTimestamp();
+        }
+
         // First we collect threads that have trace context to increase the chance to get
         CodeHotspotIteration();
         // Then we collect threads that do not have trace context
         WalltimeProfilingIteration();
+
+        if (_areInternalMetricsEnabled)
+        {
+            timestampNanosecs2 = OpSysTools::GetHighPrecisionTimestamp();
+            _walltimeDurationMetric->Add(static_cast<double>(timestampNanosecs2 - timestampNanosecs1));
+        }
     }
 
     // When CPU profiling is enabled, most of the threads are scanned
     // and if they are currently running, they are sampled.
-    if (_pConfiguration->IsCpuProfilingEnabled())
+    if (_isCpuEnabled)
     {
         CpuProfilingIteration();
+
+        if (_areInternalMetricsEnabled)
+        {
+            // avoid an unnecessary call to OpSysTools::GetHighPrecisionTimestamp() if possible
+            if (timestampNanosecs2 == 0)
+            {
+                timestampNanosecs2 = OpSysTools::GetHighPrecisionTimestamp();
+            }
+
+            timestampNanosecs1 = OpSysTools::GetHighPrecisionTimestamp();
+            _cpuDurationMetric->Add(static_cast<double>(timestampNanosecs1 - timestampNanosecs2));
+        }
```

Comment:
Here do not measure the call to `CpuProfilingIteration()`.
The value will be close to 0.
we 
Shouldn't we instead write this:
```
        if (_areInternalMetricsEnabled)
        {
            timestampNanosecs1 = OpSysTools::GetHighPrecisionTimestamp();
        }
        
        CpuProfilingIteration();
        
        if (_areInternalMetricsEnabled)
        {
            timestampNanosecs2 = OpSysTools::GetHighPrecisionTimestamp();
            _cpuDurationMetric->Add(static_cast<double>(timestampNanosecs2 - timestampNanosecs1));
        }
```